### DISCUSSION TITLE
[Native DLL] Pointer to array of strings

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#3012] [General]
C declaration: `int Init2(int *argc, char ***argv)`.

My PascalABC.NET translation: `function Init2(argc: IntPtr; argv: IntPtr): integer; external 'mydll.dll' name 'Init2';`

Currently, I call it like this: `Init2(new IntPtr(0), new IntPtr(0));`

But, I don't know if my translation is right or not, too. I also don't know how to pass the value of `argc` and `argv` since they are all pointers.

**Comment by SunSerega** [comment]
You'll need a lot of calls to methods of `System.Runtime.InteropServices.Marshal`, to allocate memory and convert data.
First, make sure you understand what memory representation `Init2` expects. For instance, how exactly does it know the number of items in `argc` array? Prob the last item is 0, but it might also look at `argv`.

**Reply by ghost** [reply]
Please show me code. Btw, this is also the first time I have seen code like that. Pointer to array of strings seems to be weird.

**Reply by SunSerega** [reply]
It's just an `array of array of string`. Or `array of array of array of char`. Another layer of indirection (of pointing) - another layer of array.

**Reply by SunSerega** [reply]
But I'm pretty sure a standard marshaler wouldn't handle jagged arrays, that's why I'm telling you to learn to use the `Marshal` class manually. It's not that hard.

The hard part is avoiding memory leaks, especially if the `external` function you call has a chance to fail. There is basically no silver bullet, you'll need to study the `external` function in question, and how it handles the memory allocation you've passed to it.
